src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -157,42 +164,59 @@ EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_sh
-		MonoBundledAssembly *cur;
-		if (XA_UNLIKELY (assembly_count < 0 || (bundled_assemblies_slow_path && bundled_assemblies.size () <= bundled_assembly_index))) {
-			if (assembly_count == -1) {
+		if (XA_UNLIKELY (bundled_assembly_index >= application_config.number_of_assemblies_in_apk || bundled_assemblies_slow_path)) {
+			if (!bundled_assemblies_slow_path && bundled_assembly_index == application_config.number_of_assemblies_in_apk) {
 				log_warn (LOG_ASSEMBLY, "Number of assemblies stored at build time (%u) was incorrect, switching to slow bundling path.");
 			}
-			bundled_assemblies.emplace_back ();
-			cur = &bundled_assemblies.back ();
-		} else {
-			cur = &bundled_assemblies [bundled_assembly_index];
-		}
-		bundled_assembly_index++;
 
-		md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
-		cur->name = utils.strdup_new (entry_name.get () + prefix_len, entry_name.length () - prefix_len);
-		cur->data = (const unsigned char*)map_info.area;
+			if (extra_bundled_assemblies == nullptr) {
+				extra_bundled_assemblies = new std::vector<XamarinAndroidBundledAssembly> ();
+			}
 
-		// MonoBundledAssembly::size is const?!
-		unsigned int *psize = (unsigned int*) &cur->size;
-		*psize = static_cast<unsigned int>(file_size);
+			extra_bundled_assemblies->emplace_back ();
+			// <true> means we need to allocate memory to store the entry name, only the entries pre-allocated during
+			// build have valid pointer to the name storage area
+			set_entry_data<true> (extra_bundled_assemblies->back (), fd, data_offset, file_size, prefix_len, max_assembly_name_size, entry_name);
+			continue;
+		}
 
-		if (XA_UNLIKELY (utils.should_log (LOG_ASSEMBLY))) {
-			const char *p = (const char*) cur->data;
+		set_assembly_entry_data (bundled_assemblies [bundled_assembly_index], fd, data_offset, file_size, prefix_len, max_assembly_name_size, entry_name);
+		bundled_assembly_index++;
+	}
 
-			std::array<char, 9> header;
-			for (size_t j = 0; j < header.size () - 1; ++j)
-				header[j] = isprint (p [j]) ? p [j] : '.';
-			header [header.size () - 1] = '\0';
+	have_and_want_debug_symbols = register_debug_symbols && bundled_debug_data != nullptr;
+}
 
-			log_info_nocheck (LOG_ASSEMBLY, "file-offset: % 8x  start: %08p  end: %08p  len: % 12i  zip-entry:  %s name: %s [%s]",
-			                  (int) data_offset, cur->data, cur->data + *psize, (int) file_size, entry_name.get (), cur->name, header.data ());
-		}
+template<bool NeedsNameAlloc>
+force_inline void
+EmbeddedAssemblies::set_entry_data (XamarinAndroidBundledAssembly &entry, int apk_fd, uint32_t data_offset, uint32_t data_size, uint32_t prefix_len, uint32_t max_name_size, dynamic_local_string<SENSIBLE_PATH_MAX> const& entry_name) noexcept
+{
+	entry.apk_fd = apk_fd;
+	if constexpr (NeedsNameAlloc) {
+		entry.name = utils.strdup_new (entry_name.get () + prefix_len);
+	} else {
+		strncpy (entry.name, entry_name.get () + prefix_len, max_name_size);
```

Comment:
Should this also explicitly set `entry.name[max_name_size]=0` or `entry.name[max_name_size-1]=0`?

Given the `+1`/`-1` operations elsewhere, I *think* `libxamarin-app.so` will allocate the `XamarinAndroidBundledAssembly::name` entry to be of length `char[max_name_size+1]` (due to the `-1` in `libmonodroid.so`, and thus it'll come "pre-NUL-terminated", but it's taken me more mental effort than I'd like to admit to come to this conclusion.

Reply:
`strncpy` copies **at most** `n` bytes (in this case `max_name_size`) which won't include the terminating NUL byte, and since this line is used only with pre-allocated buffers, the string is terminated properly even if the name takes the entire `max_name_size` characters. `strncpy` also stops copying at the first `\0` it encounters and copies it to the destination buffer.